### PR TITLE
Add breakdown feature

### PR DESCRIPTION
**Pull request by manuthecoder**
## **Description:**
This pull request includes the addition of a "Breakdown" button in the sidebar, providing users with comprehensive insights into project expenses. When clicked, this button opens a new page dedicated to showcasing various expense breakdown metrics, aiding users in better understanding their financial data.

## **New Features:**
- **Breakdown Button:** Added a new button in the sidebar for accessing expense breakdown details.
- **Breakdown Page:** Created a new page layout to display detailed expense breakdown information. Hover over a chart to export it as a PNG image. 

## **Expense Breakdown Metrics:**
- **Overall Balance by Date:** Displays the overall balance of expenses over different time periods.
- **Overall Change by Date:** Presents the overall change in expenses over different time periods.
- **Max Positive Change:** Highlights the maximum positive change in expenses.
- **Max Negative Change:** Highlights the maximum negative change in expenses.
- **Expenses by Tag:** Categorizes expenses based on predefined tags for easier tracking.
- **Expenses by User:** Segregates expenses based on the user who incurred them.
- **Expenses by Merchant:** Organizes expenses based on the merchant where the transaction occurred.
- **Expenses by Category:** Groups expenses based on predefined categories for better understanding and analysis.

## **Screenshots**
127 0 0 1_3000_hack_the_seas_breakdown

## **Important note!**
I've added fake data which appears _only on development_. Remove this before deploying to production

### CODE REVIEW

**Comment by sampoder** [comment]
WOW! Awesome. I'll check this out later this week! (Midterm season sorry)

**Comment by garyhtou** [comment]
Sooo cool!

File: app/assets/stylesheets/_dark.scss [hunk 1]
```diff
@@ -376,4 +377,8 @@ html[data-dark='true'] {
       @include gradient(info);
     }
   }
+
+  .card-breakdown {
+    border-color: map-get($palette, slate);
+  }
```
Comment (sampoder) [inline app/assets/stylesheets/_dark.scss]: Could you instead use these classes: `border b--muted` on the card?

File: app/assets/stylesheets/components/_cards.scss [hunk 2]
```diff
@@ -140,6 +151,14 @@ a.card {
   }
 }
 
+.card-breakdown {
+  border-radius: 25px;
+  padding: 20px;
+  padding-top: 0px;
+  padding-bottom: 10px;
+  border: 1px solid map-get($palette, smoke);
+}
```
Comment (sampoder) [inline app/assets/stylesheets/components/_cards.scss]: Let's use the standard card design with the same border-radius

File: app/controllers/events_controller.rb [hunk 3]
```diff
@@ -182,6 +182,135 @@ def show
...
+    @transactions_by_tag = {}
+
+    initial_subtotal = 0
+    running_total = 0
+
+    ((1.year.ago - (1.year.ago.to_date.wday % 7).day).to_date...0.day.ago.to_date).each do |date|
+      @change_by_date[date.to_s] = {
+        neg: 0,
+        pos: 0
+      }
```
Comment (sampoder) [inline app/controllers/events_controller.rb]: Can you work through making the changes in this file to make them more readable? This code I originally wrote was janky and was more of a proof of concept.

File: app/models/organizer_position.rb [hunk 4]
```diff
@@ -51,4 +51,12 @@ def tourable_options
...
   end
 
+  def spent
+    card = event.canonical_transactions.stripe_transaction.joins("JOIN stripe_cardholders on raw_stripe_transactions.stripe_transaction->>'cardholder' = stripe_cardholders.stripe_id").where(stripe_cardholders: { user_id: user.id }).sum(:amount_cents) + event.emburse_transactions.joins("JOIN emburse_cards on emburse_transactions.emburse_card_id = emburse_cards.id").where(emburse_cards: { user_id: user.id }).sum(:amount)
+    ach = event.ach_transfers.where(creator_id: user.id, rejected_at: nil).sum(:amount)
+    checks = event.checks.where(creator_id: user.id, rejected_at: nil).sum(:amount) + event.increase_checks.where(user_id: user.id, increase_status: "deposited").where.not(approved_at: nil).sum(:amount)
+    disbursements = event.disbursements.where(requested_by_id: user.id).where.not(fulfilled_by_id: nil).sum(:amount)
+    card.abs + ach.abs + checks.abs + disbursements.abs
+  end
+
```
Comment (sampoder) [inline app/models/organizer_position.rb]: Same comment as above, let's make this more readable!

File: app/views/events/breakdown.html.erb [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+  }
+  .change-column {
+    flex: 1;
+  }
+  @media only screen and (max-width: 600px) {
+    .breakdown-row {
+      flex-direction: column;
+    }
+  }
+</style>
```
Comment (sampoder) [inline app/views/events/breakdown.html.erb]: Let's put this in a dedicated file

**Review by sampoder — CHANGES_REQUESTED** [review]
This is looking really good! Just dropped some comments about what we need to do before this gets merged, feel free to DM me any questions you've got about the code :D

**Comment by manu-gurudath** [comment]
idk what the last commit did i just clicked on a button 😭

**Comment by manu-gurudath** [comment]
I just made changes based on the code you reviewed. Let me know if anything else needs to be changed! 🚀

**Review by sampoder — DISMISSED** [review]
I'm really liking the look of this & would be happy to merge as-is. I've added this cool GitHub heatmap. This is the current state:

image

My only ask is that we feature flag this and collect some user feedback before showing this to everyone. We'll probably want to make some changes.

Let's also get some other engineers' eyes on this.

**Comment by manu-gurudath** [comment]
hi! I think It would look way more aesthetically pleasing if you added the cards back instead of leaving the charts without a background color/outline

**Comment by manu-gurudath** [comment]
also if you could add a heading/label to the heatmap which said "expenses in the past year" it would look more intuitive

**Comment by manu-gurudath** [comment]
also I noticed the chart labels getting cut off
image

**Comment by jaspermayone** [comment]
So cool, happy to test this!

**Comment by manu-gurudath** [comment]
not sure what happened here...
image

File: app/models/tag.rb [hunk 6]
```diff
@@ -41,4 +41,8 @@ def removal_confirmation_message
     end
   end
 
+  def total_amount
+   
+  end
+
```
Comment (sampoder) [inline app/models/tag.rb]:
```suggestion
```

File: app/views/events/_nav.html.erb [hunk 7]
```diff
@@ -14,7 +14,12 @@
       tooltip: "View your transactions",
       icon: "transactions",
       selected: local_assigns[:selected] == :transactions %>
-    <h4 class="mb0 h4 font-medium muted">Receive</h4>
+    <%= dock_item "Breakdown",
```
Comment (sampoder) [inline app/views/events/_nav.html.erb]: Fill Flipper flag

File: app/assets/stylesheets/_utilities.scss
Comment (garyhtou) [inline app/assets/stylesheets/_utilities.scss:1]: Move into breakdown component css file

File: app/assets/stylesheets/components/_breakdown.scss [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+}
+
+.grid-item {
+  border: 0.5px solid #000;
+  border-radius: 3px;
+  padding-left: 0px!important;
+  padding-right: 0px!important;
+  padding-bottom: 0px!important;
+  padding-top: calc(100% - 1px)!important;
+}
```
Comment (garyhtou) [inline app/assets/stylesheets/components/_breakdown.scss]:
```suggestion
}

```

File: app/assets/stylesheets/components/_cards.scss [hunk 9]
```diff
@@ -155,3 +155,12 @@ a.card {
 .card--sunset {
   background-image: url('https://cloud-awvmmj0xe.vercel.app/sunset_card.png');
 }
+
+.card-breakdown {
```
Comment (garyhtou) [inline app/assets/stylesheets/components/_cards.scss]: maybe move into breakdown component?

```suggestion
.card--breakdown {
```

File: app/controllers/events_controller.rb [hunk 10]
```diff
@@ -215,6 +215,26 @@ def show
     end
   end
 
+  def breakdown
+    begin
+      authorize @event
+    rescue Pundit::NotAuthorizedError
+      return redirect_to root_path, flash: { error: "We couldn’t find that organization!" }
+    end
```
Comment (garyhtou) [inline app/controllers/events_controller.rb]:
```suggestion
    authorize @event
```

File: app/controllers/events_controller.rb [hunk 11]
```diff
@@ -215,6 +215,26 @@ def show
...
+
+    @heatmap = BreakdownEngine::Heatmap.new(@event).run
+    @maximum_positive_change = @heatmap.values.map { |change| change[:positive] }.max || 0
+    @maximum_negative_change = @heatmap.values.map { |change| change[:negative] }.min || 0
```
Comment (garyhtou) [inline app/controllers/events_controller.rb]:
```suggestion
    @maximum_positive_change = @heatmap.values.max { |change| change[:positive] } || 0
    @maximum_negative_change = @heatmap.values.min { |change| change[:negative] } || 0
```

File: app/policies/event_policy.rb [hunk 12]
```diff
@@ -21,6 +21,10 @@ def show?
     is_public || admin_or_user?
   end
 
+  def breakdown?
+    is_public || admin_or_user?
```
Comment (garyhtou) [inline app/policies/event_policy.rb]: For the time being, let's not making it transparent yet

```suggestion
    admin_or_user?
```

File: app/services/breakdown_engine/categories.rb [hunk 13]
```diff
@@ -0,0 +1,24 @@
...
+    def initialize(event)
+      @event = event
+    end
+
+    def run
+      RawStripeTransaction.select(
+        "trim(upper(split_part(raw_stripe_transactions.stripe_transaction->'merchant_data'->>'category', '*', 1))) AS category",
+        "SUM(raw_stripe_transactions.amount_cents) * -1 AS amount_cents"
+      )
+                          .joins("LEFT JOIN canonical_transactions ct ON raw_stripe_transactions.id = ct.transaction_source_id")
```
Comment (garyhtou) [inline app/services/breakdown_engine/categories.rb]:
```suggestion
                          .joins("LEFT JOIN canonical_transactions ct ON raw_stripe_transactions.id = ct.transaction_source_id and ct.transaction_source_type = ${RawStripeTransaction.class}")
```

File: app/services/breakdown_engine/heatmap.rb [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+      end
+
+      last_date = heatmap.keys.max.to_date
+
+      ((last_date + 1.day).to_date..6.days.from_now.to_date).each do |date|
+        heatmap[date.to_s] = {
+          negative: 0,
+          positive: 0
+        }
+      end
```
Comment (garyhtou) [inline app/services/breakdown_engine/heatmap.rb]:
```suggestion
      last_date = heatmap.keys.max.to_date

      ((last_date + 1.day).to_date..6.days.from_now.to_date).each do |date|
        heatmap[date.to_s] = {
          negative: 0,
          positive: 0
        }
      end
```

File: app/services/breakdown_engine/heatmap.rb [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+    def initialize(event)
+      @event = event
+    end
+
+    def run
+      heatmap = {}
+
+      start = 1.year.ago - (1.year.ago.to_date.wday % 7).day
+
+      range = start.to_date...0.days.ago.to_date
```
Comment (garyhtou) [inline app/services/breakdown_engine/heatmap.rb]:
```suggestion
      range = start.to_date...6.days.from_now.to_date
```

File: app/services/breakdown_engine/merchants.rb [hunk 16]
```diff
@@ -0,0 +1,28 @@
...
+    def initialize(event)
+      @event = event
+    end
+
+    def run
+      RawStripeTransaction.select(
+        "CASE
+           WHEN raw_stripe_transactions.stripe_transaction->'merchant_data'->>'name' SIMILAR TO '(SQ|GOOGLE|TST|RAZ|INF|PayUp|IN|INT|\\*)%'
+             THEN TRIM(UPPER(raw_stripe_transactions.stripe_transaction->'merchant_data'->>'name'))
+           ELSE TRIM(UPPER(SPLIT_PART(raw_stripe_transactions.stripe_transaction->'merchant_data'->>'name', '*', 1)))
```
Comment (garyhtou) [inline app/services/breakdown_engine/merchants.rb]: do the `THEN` and `ELSE` need to be switched?

File: app/services/breakdown_engine/merchants.rb [hunk 17]
```diff
@@ -0,0 +1,28 @@
...
+         END AS merchant",
+        "SUM(raw_stripe_transactions.amount_cents) * -1 AS amount_cents"
+      )
+                          .joins("LEFT JOIN canonical_transactions ct ON raw_stripe_transactions.id = ct.transaction_source_id")
```
Comment (garyhtou) [inline app/services/breakdown_engine/merchants.rb]: `transaction_source_type` problem from above

File: app/services/breakdown_engine/tags.rb [hunk 18]
```diff
@@ -0,0 +1,18 @@
+# frozen_string_literal: true
+
+module BreakdownEngine
+  class Tags
+    def initialize(event)
+      @event = event
+    end
+
+    def run
+      tags = @event.tags.each_with_object({}) do |tag, hash|
```
Comment (garyhtou) [inline app/services/breakdown_engine/tags.rb]:
```suggestion
      tags = @event.tags.includes(hcb_codes: [:canonical_transactions, :canonical_pending_transactions]).each_with_object({}) do |tag, hash|
```

File: app/services/breakdown_engine/tags.rb [hunk 19]
```diff
@@ -0,0 +1,18 @@
...
+        hash[tag.label] = tag.hcb_codes.select{ |hcb_code| hcb_code.amount_cents < 0 }.sum { |hcb_code| hcb_code.amount_cents }.to_f / 100
+      end
+
+      tags.select { |key, value| value > 0 }
```
Comment (garyhtou) [inline app/services/breakdown_engine/tags.rb]:
```suggestion
      tags = @event.tags.includes(hcb_codes: [:canonical_transactions, :canonical_pending_transactions]).each_with_object({}) do |tag, hash|
        amount_cents_sum = tag.hcb_codes.sum do |hcb_code|
          [hcb_code.amount_cents, 0].min
        end
        hash[tag.label] = (amount_cents_sum * -1).to_f / 100 if amount_cents_sum > 0
      end
```

File: app/services/breakdown_engine/users.rb [hunk 20]
```diff
@@ -0,0 +1,23 @@
...
+    def run
+      @event.organizer_positions.includes(:user)
+            .each_with_object({}) { |position, hash|
+        hash[position.user.name] = @event.canonical_transactions
+                                         .stripe_transaction
+                                         .joins("JOIN stripe_cardholders ON raw_stripe_transactions.stripe_transaction->>'cardholder' = stripe_cardholders.stripe_id")
+                                         .where(stripe_cardholders: {
+                                                  user_id: position.user.id
+                                                })
+                                         .sum(:amount_cents).abs.to_f / 100
```
Comment (garyhtou) [inline app/services/breakdown_engine/users.rb]:
```suggestion
                                         .sum(:amount_cents).to_f / 100 * -1
```

**Review by garyhtou — COMMENTED** [review]
let's feature flag & merge!

**Review by garyhtou — COMMENTED** [review]
let's feature flag & merge!

**Pull request merged**